src/swap/handle_swap_sign_transaction.c

```diff
@@ -136,79 +206,73 @@ bool swap_check_validity() {
         os_sched_exit(0);
     }
 
-    if (G_context.tx_info.transaction.hints_type != TRANSACTION_COMMENT) {
+    if ((G_context.tx_info.transaction.hints_type == TRANSACTION_COMMENT) &&
+        (G_context.tx_info.transaction.hints_len != 0)) {
+        PRINTF("Hint length %d refused\n", G_context.tx_info.transaction.hints_len);
+        io_send_sw(SW_SWAP_FAILURE);
+        // unreachable
+        os_sched_exit(0);
+    } else if ((G_context.tx_info.transaction.hints_type != TRANSACTION_COMMENT) &&
+               (G_context.tx_info.transaction.hints_type != TRANSACTION_TRANSFER_JETTON)) {
         PRINTF("Wrong operation %d\n", G_context.tx_info.transaction.hints_type);
         io_send_sw(SW_SWAP_FAILURE);
         // unreachable
         os_sched_exit(0);
-    } else if (G_context.tx_info.transaction.hints_len != 0) {
-        PRINTF("Hint length %d refused\n", G_context.tx_info.transaction.hints_len);
+    }
+
+    if ((strncmp(G_swap_validated.ticker, "TON", sizeof("TON")) == 0) &&
+        (G_context.tx_info.transaction.hints_type == TRANSACTION_TRANSFER_JETTON)) {
+        PRINTF("Wrong operation type %d for native coin swap\n",
+               G_context.tx_info.transaction.hints_type);
```

Comment:
I think we should also check the opposite case, where we receive a `TRANSACTION_COMMENT` but `G_swap_validated.ticker != "TON"`